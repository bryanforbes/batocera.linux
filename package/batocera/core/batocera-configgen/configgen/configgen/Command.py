from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence
    from pathlib import Path


class Command:
    def __init__(self, array: Sequence[str | Path], env: Mapping[str, str | Path] = {}):
        self.array = list(array)
        self.env = dict(env)

    def __str__(self):
        strings = [f"{varName}={varValue}" for varName, varValue in self.env.items()]

        strings.extend(str(value) for value in self.array)

        return " ".join(strings)
