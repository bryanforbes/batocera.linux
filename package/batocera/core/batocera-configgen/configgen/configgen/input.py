from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass, replace
from typing import TYPE_CHECKING, Self, TypeAlias, TypedDict, Unpack, cast

if TYPE_CHECKING:
    import xml.etree.ElementTree as ET


class _InputChanges(TypedDict, total=False):
    name: str
    type: str
    id: str
    value: str
    code: str


@dataclass(slots=True, kw_only=True)
class Input:
    name: str
    type: str
    id: str
    value: str
    code: str

    def replace(self, /, **changes: Unpack[_InputChanges]) -> Self:
        return replace(self, **changes)

    @classmethod
    def from_element(cls, element: ET.Element, /) -> Self:
        return cls(
            name=cast(str, element.get("name")),
            type=cast(str, element.get("type")),
            id=cast(str, element.get("id")),
            value=cast(str, element.get("value")),
            code=cast(str, element.get("code"))
        )

    @classmethod
    def from_parent_element(cls, parent_element: ET.Element, /) -> dict[str, Self]:
        inputs: dict[str, Self] = {}

        for element in parent_element.findall('input'):
            input = cls.from_element(element)
            inputs[input.name] = input

        return inputs

InputMapping: TypeAlias = Mapping[str, Input]
InputDict: TypeAlias = dict[str, Input]
