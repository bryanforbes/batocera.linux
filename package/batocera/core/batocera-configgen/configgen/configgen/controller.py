from __future__ import annotations

import xml.etree.ElementTree as ET
from collections.abc import Mapping
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import TYPE_CHECKING, Final, Literal, Self, TypeAlias, TypedDict, Unpack, cast

from .batoceraPaths import BATOCERA_ES_DIR, USER_ES_DIR
from .input import Input, InputDict

if TYPE_CHECKING:
    import argparse

"""Default mapping of Batocera keys to SDL_GAMECONTROLLERCONFIG keys."""
_DEFAULT_SDL_MAPPING: Final = {
    'b': 'a',
    'a': 'b',
    'x': 'y',
    'y': 'x',
    'l2': 'lefttrigger',
    'r2': 'righttrigger',
    'l3': 'leftstick',
    'r3': 'rightstick',
    'pageup': 'leftshoulder',
    'pagedown': 'rightshoulder',
    'start': 'start',
    'select': 'back',
    'up': 'dpup',
    'down': 'dpdown',
    'left': 'dpleft',
    'right': 'dpright',
    'joystick1up': 'lefty',
    'joystick1left': 'leftx',
    'joystick2up': 'righty',
    'joystick2left': 'rightx',
    'hotkey': 'guide'
}


def _key_to_sdl_game_controller_config(keyname: str, name: str, type: str, id: str, value: str) -> str | None:
    """
    Converts a key mapping to the SDL_GAMECONTROLLER format.

    Arguments:
      keyname: (str) SDL_GAMECONTROLLERCONFIG input name.
      name: (str) `es_input.cfg` input name.
      type: (str) 'button', 'hat', or 'axis'
      id: (int) Numeric key id.
      value: (int) Hat value. Only used if type == 'hat' or type == 'axis' and 'joystick' in name.
    Returns:
      (str) SDL_GAMECONTROLLERCONFIG-formatted key mapping string.
    Examples:
      _keyToSdlGameControllerConfig('leftshoulder', 'l1', 'button', 6)
        'leftshoulder:b6'

      _keyToSdlGameControllerConfig('dpleft', 'left', 'hat', 0, 8)
        'dpleft:h0.8'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, -1)
        'lefty:a1'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, 1)
        'lefty:a1~'

      _keyToSdlGameControllerConfig('dpup', 'up', 'axis', 1, -1)
        'dpup:-a1'
    """
    if type == 'button':
        return f'{keyname}:b{id}'
    elif type == 'hat':
        return f'{keyname}:h{id}.{value}'
    elif type == 'axis':
        if 'joystick' in name:
            return f"{keyname}:a{id}{'~' if int(value) > 0 else ''}"
        elif keyname in ('dpup', 'dpdown', 'dpleft', 'dpright'):
            return f"{keyname}:{'-' if int(value) < 0 else '+'}a{id}"
        else:
            return f'{keyname}:a{id}'
    elif type == 'key':
        return None
    else:
        raise ValueError(f'unknown key type: {type!r}')


class _ControllerChanges(TypedDict, total=False):
    guid: str
    player: int | None
    index: int
    real_name: str
    device_path: str | None
    button_count: int | None
    hat_count: int | None
    axis_count: int | None
    physical_device_path: str | None
    physical_id: int | None


@dataclass(slots=True, kw_only=True)
class Controller:
    device_name: str
    type: Literal['keyboard', 'joystick']
    guid: str
    player: int | None
    index: int = -1
    real_name: str = ""
    inputs: InputDict = field(default_factory=dict)
    device_path: str | None = None
    button_count: int | None = None
    hat_count: int | None = None
    axis_count: int | None = None
    physical_device_path: str | None = None
    physical_id: int | None = None

    @property
    def uid_name(self) -> str:
        return self.guid + self.device_name

    def replace(self, /, **changes: Unpack[_ControllerChanges]) -> Self:
        return replace(self, **changes)

    def generate_sdl_game_db_line(self, sdl_mapping: Mapping[str, str] = _DEFAULT_SDL_MAPPING, /) -> str:
        config = [self.guid, self.real_name, 'platform:Linux']

        def add_mapping(input: Input) -> None:
            key_name = sdl_mapping.get(input.name, None)
            if key_name is None:
                return
            sdl_config = _key_to_sdl_game_controller_config(key_name, input.name, input.type, input.id, input.value)
            if sdl_config is not None:
                config.append(sdl_config)

        hotkey_input: Input | None = None
        mapped_button_ids: set[str] = set()
        for input in self.inputs.values():
            if input.name is None:
                continue
            if input.name == 'hotkey':
                hotkey_input = input
                continue
            if input.type == 'button':
                mapped_button_ids.add(input.id)

            add_mapping(input)

        if hotkey_input is not None and hotkey_input.id not in mapped_button_ids:
            add_mapping(hotkey_input)

        config.append('')

        return ','.join(config)

    @classmethod
    def from_element(cls, element: ET.Element, /) -> Self:
        return cls(
            device_name=cast(str, element.get("deviceName")),
            type=cast(Literal['keyboard', 'joystick'], element.get("type")),
            guid=cast(str, element.get("deviceGUID")),
            player=None,
            index=-1,
            inputs=Input.from_parent_element(element)
        )

    @classmethod
    def parse_all(cls) -> list[Self]:
        return [
            cls.from_element(controller)
            for conffile in [BATOCERA_ES_DIR / "es_input.cfg", USER_ES_DIR / 'es_input.cfg'] if conffile.exists()
            for controller in ET.parse(conffile).getroot().findall(".//inputConfig")
        ]

    @classmethod
    def load_all(cls) -> dict[str, Self]:
        return {controller.uid_name: controller for controller in cls.parse_all()}

    @classmethod
    def load_for_players(cls, max_players: int, args: argparse.Namespace, /) -> dict[int, Self]:
        controllers: dict[int, Self] = {}
        all_controllers = cls.load_all()

        for player_number in range(1, max_players + 1):
            controller = cls.find_best_controller_config(all_controllers, args, player_number)
            if controller is not None:
                controllers[player_number] = controller

        return controllers

    @classmethod
    def find_best_controller_config(
        cls, controllers: Mapping[str, Self], args: argparse.Namespace, player_number: int, /,
    ) -> Self | None:
        player_index: int | None = getattr(args, f'p{player_number}index')

        if player_index is None:
            return None

        player_guid: str = getattr(args, f'p{player_number}guid')
        player_name: str = getattr(args, f'p{player_number}name')
        player_device_path: str = getattr(args, f'p{player_number}devicepath')
        player_buttons: int = getattr(args, f'p{player_number}nbbuttons')
        player_hats: int = getattr(args, f'p{player_number}nbhats')
        player_axes: int = getattr(args, f'p{player_number}nbaxes')

        for controller in controllers.values():
            if controller.guid == player_guid and controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.guid == player_guid:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        return None


ControllerMapping: TypeAlias = Mapping[str, Controller]
ControllerDict: TypeAlias = Mapping[str, Controller]
ControllerPlayerMapping: TypeAlias = Mapping[int, Controller]
ControllerPlayerDict: TypeAlias = dict[int, Controller]


def generate_sdl_game_controller_config(controllers: ControllerPlayerMapping, /) -> str:
    return '\n'.join([controller.generate_sdl_game_db_line() for controller in controllers.values()])

def write_sdl_controller_db(
    controllers: ControllerPlayerMapping, output_file: str | Path = '/tmp/gamecontrollerdb.txt', /,
) -> Path:
    output_file = Path(output_file)
    with output_file.open('w') as text_file:
        text_file.write(generate_sdl_game_controller_config(controllers))
    return output_file
