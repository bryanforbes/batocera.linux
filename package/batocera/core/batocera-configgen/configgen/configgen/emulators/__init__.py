from __future__ import annotations

from contextlib import nullcontext
from importlib import import_module
from typing import TYPE_CHECKING, Final

from ..batoceraPaths import BATOCERA_CONF
from ..config import RenderConfig, SystemConfig
from ..controller import Controller
from ..controllersConfig import getGamesMetaData
from ..gun import Gun, precalibrate_guns
from ..settings.unixSettings import UnixSettings
from ..utils.squashfs import squashfs_rom
from ..utils.wheels import configure_wheels
from ._base import Emulator as Emulator  # noqa: TCH001

if TYPE_CHECKING:
    from ..cli import Options

_GENERATOR_MAP: Final[dict[str, str]] = {}


def get_emulator_cls(emulator: str, /) -> type[Emulator]:
    module_path = _GENERATOR_MAP.get(emulator, emulator.replace('-', '_'))

    try:
        module = import_module(f'..{module_path}', package=__name__)
        emulator_cls: type[Emulator] = module.Emulator
    except ImportError as e:
        if e.name is not None and e.name.startswith(__name__.split('.')[0]):
            raise Exception(f'no emulator found for {emulator!r}') from e

        raise
    except AttributeError as e:
        raise Exception(f'no emulator found for {emulator!r}') from e

    return emulator_cls


def run_emulator(options: Options, /) -> int:
    exit_code = -1
    original_rom = options['rom']

    if original_rom.suffix == '.squashfs':
        rom_context_manager = squashfs_rom(original_rom)
    else:
        rom_context_manager = nullcontext(original_rom)

    with rom_context_manager as rom:
        system = options['system']

        settings = UnixSettings(BATOCERA_CONF)
        system_config = SystemConfig.load(options, settings)
        render_config = RenderConfig.load(options, settings, system_config.get('shaderset'))

        metadata = getGamesMetaData(system, original_rom)

        controllers = Controller.load_for_player_options(options['players'])

        if system_config.use_guns:
            guns = Gun.get_all()
            precalibrate_guns(system, system_config.emulator, system_config.core, original_rom)
        else:
            guns = {}

        with configure_wheels(controllers, system, system_config, metadata) as (controllers, wheels):
            emulator_cls = get_emulator_cls(system_config.emulator)
            emulator = emulator_cls(system, system_config, render_config, metadata, controllers, guns, wheels)
            exit_code = emulator.run(rom, original_rom, options['gameinfoxml'])

    return exit_code
