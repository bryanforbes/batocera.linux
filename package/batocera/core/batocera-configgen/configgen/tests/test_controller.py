from __future__ import annotations

from argparse import Namespace
from typing import TYPE_CHECKING

import pytest

from configgen.batoceraPaths import BATOCERA_ES_DIR, HOME, USER_ES_DIR
from configgen.controller import Controller
from configgen.exceptions import BatoceraException
from configgen.input import Input
from tests.mock_controllers import make_player_controller

if TYPE_CHECKING:
    from unittest.mock import Mock

    from pyfakefs.fake_filesystem import FakeFilesystem
    from pytest_mock.plugin import MockerFixture
    from syrupy.assertion import SnapshotAssertion


@pytest.fixture(autouse=True)
def evdev(evdev: Mock) -> Mock:
    evdev.ecodes.EV_ABS = 3
    evdev.ecodes.ABS_HAT0X = 0x10
    return evdev


class TestController:
    def test_init(self, snapshot: SnapshotAssertion) -> None:
        input = Input(name='a', type='button', id='1', value='1', code='305')
        inputs = {input.name: input}
        controller = Controller(
            name='name',
            type='joystick',
            guid='guid',
            player_number=1,
            index=0,
            real_name='real name',
            device_path='/dev/null',
            button_count=1,
            hat_count=2,
            axis_count=3,
            inputs_=inputs,
        )
        assert controller == snapshot
        assert controller.inputs is not inputs

    def test_generate_sdl_game_db_line(self, snapshot: SnapshotAssertion) -> None:
        controller = Controller(
            name='Xbox 360 Wireless Receiver (XBOX)',
            type='joystick',
            guid='030000005e040000a102000007010000',
            player_number=1,
            index=0,
            real_name='real name',
            device_path='/dev/null',
            button_count=1,
            hat_count=2,
            axis_count=3,
            inputs_={
                'a': Input(name='a', type='button', id='1', value='1', code='305'),
                'b': Input(name='b', type='button', id='0', value='1', code='304'),
                'down': Input(name='down', type='button', id='16', value='1', code='707'),
                'right': Input(name='right', type='axis', id='0', value='1', code='0'),
                'left': Input(name='left', type='axis', id='0', value='-1', code='0'),
                'up': Input(name='up', type='hat', id='0', value='1', code='16'),
                'hotkey': Input(name='hotkey', type='button', id='10', value='1', code='316'),
                'joystick1left': Input(name='joystick1left', type='axis', id='0', value='-1', code='0'),
                'joystick1up': Input(name='joystick1up', type='axis', id='1', value='-1', code='1'),
                'joystick2up': Input(name='joystick2up', type='axis', id='3', value='1', code='5'),
                'l2': Input(name='l2', type='axis', id='2', value='1', code='2'),
                'pagedown': Input(name='pagedown', type='key', id='45', value='1', code=None),
                'foo': Input(name='foo', type='key', id='45', value='1', code=None),
            },
        )

        assert controller.generate_sdl_game_db_line() == snapshot

    def test_generate_sdl_game_db_line_ignore_buttons(self, snapshot: SnapshotAssertion) -> None:
        controller = Controller(
            name='Xbox 360 Wireless Receiver (XBOX)',
            type='joystick',
            guid='030000005e040000a102000007010000',
            player_number=1,
            index=0,
            real_name='real name',
            device_path='/dev/null',
            button_count=1,
            hat_count=2,
            axis_count=3,
            inputs_={
                'a': Input(name='a', type='button', id='1', value='1', code='305'),
                'b': Input(name='b', type='button', id='0', value='1', code='304'),
                'down': Input(name='down', type='button', id='16', value='1', code='707'),
                'right': Input(name='right', type='axis', id='0', value='1', code='0'),
                'left': Input(name='left', type='axis', id='0', value='-1', code='0'),
                'up': Input(name='up', type='hat', id='0', value='1', code='16'),
                'hotkey': Input(name='hotkey', type='button', id='10', value='1', code='316'),
                'joystick1left': Input(name='joystick1left', type='axis', id='0', value='-1', code='0'),
                'joystick1up': Input(name='joystick1up', type='axis', id='1', value='-1', code='1'),
                'joystick2up': Input(name='joystick2up', type='axis', id='3', value='1', code='5'),
                'l2': Input(name='l2', type='axis', id='2', value='1', code='2'),
                'pagedown': Input(name='pagedown', type='key', id='45', value='1', code=None),
                'foo': Input(name='foo', type='key', id='45', value='1', code=None),
            },
        )

        assert controller.generate_sdl_game_db_line(ignore_buttons=['hotkey']) == snapshot

    def test_generate_sdl_game_db_line_no_hotkey(self, snapshot: SnapshotAssertion) -> None:
        controller = Controller(
            name='Xbox 360 Wireless Receiver (XBOX)',
            type='joystick',
            guid='030000005e040000a102000007010000',
            player_number=1,
            index=0,
            real_name='real name',
            device_path='/dev/null',
            button_count=1,
            hat_count=2,
            axis_count=3,
            inputs_={
                'a': Input(name='a', type='button', id='1', value='1', code='305'),
                'b': Input(name='b', type='button', id='0', value='1', code='304'),
                'down': Input(name='down', type='button', id='16', value='1', code='707'),
                'right': Input(name='right', type='axis', id='0', value='1', code='0'),
                'left': Input(name='left', type='axis', id='0', value='-1', code='0'),
                'up': Input(name='up', type='hat', id='0', value='1', code='16'),
                'joystick1left': Input(name='joystick1left', type='axis', id='0', value='-1', code='0'),
                'joystick1up': Input(name='joystick1up', type='axis', id='1', value='-1', code='1'),
                'joystick2up': Input(name='joystick2up', type='axis', id='3', value='1', code='5'),
                'l2': Input(name='l2', type='axis', id='2', value='1', code='2'),
                'pagedown': Input(name='pagedown', type='key', id='45', value='1', code=None),
                'foo': Input(name='foo', type='key', id='45', value='1', code=None),
            },
        )

        assert controller.generate_sdl_game_db_line() == snapshot

    def test_generate_sdl_game_db_line_hotkey_not_button(self, snapshot: SnapshotAssertion) -> None:
        controller = Controller(
            name='Xbox 360 Wireless Receiver (XBOX)',
            type='joystick',
            guid='030000005e040000a102000007010000',
            player_number=1,
            index=0,
            real_name='real name',
            device_path='/dev/null',
            button_count=1,
            hat_count=2,
            axis_count=3,
            inputs_={
                'a': Input(name='a', type='button', id='1', value='1', code='305'),
                'b': Input(name='b', type='button', id='0', value='1', code='304'),
                'down': Input(name='down', type='button', id='16', value='1', code='707'),
                'right': Input(name='right', type='axis', id='0', value='1', code='0'),
                'left': Input(name='left', type='axis', id='0', value='-1', code='0'),
                'up': Input(name='up', type='hat', id='0', value='1', code='16'),
                'hotkey': Input(name='hotkey', type='key', id='0', value='1', code='31546504'),
                'joystick1left': Input(name='joystick1left', type='axis', id='0', value='-1', code='0'),
                'joystick1up': Input(name='joystick1up', type='axis', id='1', value='-1', code='1'),
                'joystick2up': Input(name='joystick2up', type='axis', id='3', value='1', code='5'),
                'l2': Input(name='l2', type='axis', id='2', value='1', code='2'),
                'pagedown': Input(name='pagedown', type='key', id='45', value='1', code=None),
                'foo': Input(name='foo', type='key', id='45', value='1', code=None),
            },
        )

        assert controller.generate_sdl_game_db_line() == snapshot

    def test_generate_sdl_game_db_line_hotkey_raises(self) -> None:
        controller = Controller(
            name='Xbox 360 Wireless Receiver (XBOX)',
            type='joystick',
            guid='030000005e040000a102000007010000',
            player_number=1,
            index=0,
            real_name='real name',
            device_path='/dev/null',
            button_count=1,
            hat_count=2,
            axis_count=3,
            inputs_={
                'a': Input(name='a', type='foo', id='1', value='1', code='305'),
            },
        )

        with pytest.raises(BatoceraException, match=r"^Unknown controller input type: 'foo'"):
            controller.generate_sdl_game_db_line()

    def test_load_for_players(self, fs: FakeFilesystem, snapshot: SnapshotAssertion) -> None:
        fs.create_file(
            BATOCERA_ES_DIR / 'es_input.cfg',
            contents="""<?xml version="1.0"?>
<inputList>
	<inputConfig type="joystick" deviceName="Test Controller" deviceGUID="1">
		<input name="a" type="button" id="0" value="1" code="111" />
	</inputConfig>
	<inputConfig type="joystick" deviceName="Test Controller 2" deviceGUID="3">
		<input name="x" type="button" id="0" value="1" code="111" />
	</inputConfig>
</inputList>
""",
        )
        fs.create_file(
            USER_ES_DIR / 'es_input.cfg',
            contents="""<?xml version="1.0"?>
<inputList>
	<inputConfig type="joystick" deviceName="Test Controller" deviceGUID="2">
		<input name="b" type="button" id="0" value="1" code="222" />
	</inputConfig>
</inputList>
""",
        )

        assert (
            Controller.load_for_players(
                6,
                Namespace(
                    # P1: matches user test controller with guid and name
                    p1index=0,
                    p1guid='2',
                    p1name='Test Controller',
                    p1devicepath='/dev/event/input1',
                    p1nbbuttons=1,
                    p1nbhats=2,
                    p1nbaxes=3,
                    # P2: matches user test controller with guid
                    p2index=22,
                    p2guid='2',
                    p2name='Something else',
                    p2devicepath='/dev/event/input2',
                    p2nbbuttons=1,
                    p2nbhats=2,
                    p2nbaxes=3,
                    # P3: matches user test controller with name
                    p3index=2,
                    p3guid='20',
                    p3name='Test Controller',
                    p3devicepath='/dev/event/input3',
                    p3nbbuttons=1,
                    p3nbhats=2,
                    p3nbaxes=3,
                    # P4: matches system test controller with name and guid
                    p4index=4,
                    p4guid='1',
                    p4name='Test Controller',
                    p4devicepath='/dev/event/input4',
                    p4nbbuttons=1,
                    p4nbhats=2,
                    p4nbaxes=3,
                    # P5: matches system test controller 2 with name and guid
                    p5index=6,
                    p5guid='3',
                    p5name='Test controller 2',
                    p5devicepath='/dev/event/input6',
                    p5nbbuttons=1,
                    p5nbhats=2,
                    p5nbaxes=3,
                    # P6: not passed to emulatorlauncher
                    p6index=None,
                    p6guid=None,
                    p6name=None,
                    p6devicepath=None,
                    p6nbbuttons=None,
                    p6nbhats=None,
                    p6nbaxes=None,
                ),
            )
            == snapshot
        )

    def test_load_for_players_no_user_config(self, fs: FakeFilesystem, snapshot: SnapshotAssertion) -> None:
        fs.create_file(
            BATOCERA_ES_DIR / 'es_input.cfg',
            contents="""<?xml version="1.0"?>
<inputList>
	<inputConfig type="joystick" deviceName="Test Controller" deviceGUID="1">
		<input name="a" type="button" id="0" value="1" code="111" />
	</inputConfig>
	<inputConfig type="joystick" deviceName="Test Controller 2" deviceGUID="3">
		<input name="x" type="button" id="0" value="1" code="111" />
	</inputConfig>
</inputList>
""",
        )

        assert (
            Controller.load_for_players(
                6,
                Namespace(
                    # P1: matches test controller with guid and name
                    p1index=0,
                    p1guid='1',
                    p1name='Test Controller',
                    p1devicepath='/dev/event/input1',
                    p1nbbuttons=1,
                    p1nbhats=2,
                    p1nbaxes=3,
                    # P2: matches test controller with guid
                    p2index=22,
                    p2guid='1',
                    p2name='Something else',
                    p2devicepath='/dev/event/input2',
                    p2nbbuttons=1,
                    p2nbhats=2,
                    p2nbaxes=3,
                    # P3: matches test controller with name
                    p3index=2,
                    p3guid='20',
                    p3name='Test Controller',
                    p3devicepath='/dev/event/input3',
                    p3nbbuttons=1,
                    p3nbhats=2,
                    p3nbaxes=3,
                    # P4: matches test controller with name and guid
                    p4index=4,
                    p4guid='1',
                    p4name='Test Controller',
                    p4devicepath='/dev/event/input4',
                    p4nbbuttons=1,
                    p4nbhats=2,
                    p4nbaxes=3,
                    # P5: matches test controller 2 with name and guid
                    p5index=6,
                    p5guid='3',
                    p5name='Test controller 2',
                    p5devicepath='/dev/event/input6',
                    p5nbbuttons=1,
                    p5nbhats=2,
                    p5nbaxes=3,
                    # P6: not passed to emulatorlauncher
                    p6index=None,
                    p6guid=None,
                    p6name=None,
                    p6devicepath=None,
                    p6nbbuttons=None,
                    p6nbhats=None,
                    p6nbaxes=None,
                ),
            )
            == snapshot
        )

    @pytest.mark.usefixtures('fs')
    def test_load_for_players_no_files_raises(self) -> None:
        with pytest.raises(
            BatoceraException, match=r'^Could not find controller data for "Test Controller" with GUID "2"$'
        ):
            Controller.load_for_players(
                1,
                Namespace(
                    p1index=0,
                    p1guid='2',
                    p1name='Test Controller',
                    p1devicepath='/dev/event/input1',
                    p1nbbuttons=1,
                    p1nbhats=2,
                    p1nbaxes=3,
                ),
            )

    def test_load_for_players_raises(self, fs: FakeFilesystem) -> None:
        fs.create_file(
            BATOCERA_ES_DIR / 'es_input.cfg',
            contents="""<?xml version="1.0"?>
<inputList>
</inputList>
""",
        )

        with pytest.raises(
            BatoceraException, match=r'^Could not find controller data for "Test Controller" with GUID "2"$'
        ):
            Controller.load_for_players(
                1,
                Namespace(
                    p1index=0,
                    p1guid='2',
                    p1name='Test Controller',
                    p1devicepath='/dev/event/input1',
                    p1nbbuttons=1,
                    p1nbhats=2,
                    p1nbaxes=3,
                ),
            )

    def test_find_player_number(self, generic_xbox_pad: Controller) -> None:
        player1 = make_player_controller(generic_xbox_pad, 1)
        player2 = make_player_controller(generic_xbox_pad, 2)

        assert Controller.find_player_number([], 1) is None
        assert Controller.find_player_number([player1], 1) is player1
        assert Controller.find_player_number([player1, player2], 1) is player1
        assert Controller.find_player_number([player1, player2], 2) is player2

    @pytest.mark.usefixtures('fs')
    def test_get_mapping_axis_relaxed_values(self, generic_xbox_pad: Controller) -> None:
        assert generic_xbox_pad.get_mapping_axis_relaxed_values() == {}

    def test_get_mapping_axis_relaxed_values_with_cache_file(
        self, mocker: MockerFixture, evdev: Mock, fs: FakeFilesystem, ps3_controller: Controller
    ) -> None:
        fs.create_file(
            HOME / '.sdl2' / f'{ps3_controller.guid}_{ps3_controller.name}.cache',
            contents="""6
4000
-4000
3999
4001
-4001
-3999
""",
        )

        mock_input_device = mocker.Mock()
        mock_input_device.capabilities.return_value = {
            0: [],
            1: [],
            3: [
                (0, {}),
                (1, {}),
                (2, {}),
                (3, {}),
                (4, {}),
                (5, {}),
                (0x10, {}),
            ],
        }

        evdev.InputDevice.return_value = mock_input_device

        assert ps3_controller.get_mapping_axis_relaxed_values() == {
            'joystick1left': {'centered': False, 'reversed': False},
            'joystick1up': {'centered': False, 'reversed': False},
            'joystick2left': {'centered': False, 'reversed': True},
            'joystick2up': {'centered': False, 'reversed': False},
            'l2': {'centered': True, 'reversed': False},
            'r2': {'centered': True, 'reversed': False},
        }

    def test_get_mapping_axis_relaxed_values_with_cache_file_not_found(
        self, mocker: MockerFixture, evdev: Mock, fs: FakeFilesystem, ps3_controller: Controller
    ) -> None:
        fs.create_file(
            HOME / '.sdl2' / f'{ps3_controller.guid}_{ps3_controller.name}.cache',
            contents="""1
4001
""",
        )

        mock_input_device = mocker.Mock()
        mock_input_device.capabilities.return_value = {
            0: [],
            1: [],
            3: [
                (0x10, {}),
            ],
        }

        evdev.InputDevice.return_value = mock_input_device

        assert ps3_controller.get_mapping_axis_relaxed_values() == {
            'joystick1left': {'centered': True, 'reversed': False},
            'joystick1up': {'centered': True, 'reversed': False},
            'joystick2left': {'centered': True, 'reversed': False},
            'joystick2up': {'centered': True, 'reversed': False},
            'l2': {'centered': True, 'reversed': False},
            'r2': {'centered': True, 'reversed': False},
        }
