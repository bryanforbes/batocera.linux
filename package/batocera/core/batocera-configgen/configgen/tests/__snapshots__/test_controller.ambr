# serializer version: 1
# name: TestController.test_generate_sdl_game_db_line
  '030000005e040000a102000007010000,real name,platform:Linux,b:b1,a:b0,dpdown:b16,dpright:+a0,dpleft:-a0,dpup:h0.1,leftx:a0,lefty:a1,righty:a3~,lefttrigger:a2,guide:b10,'
# ---
# name: TestController.test_generate_sdl_game_db_line_hotkey_not_button
  '030000005e040000a102000007010000,real name,platform:Linux,b:b1,a:b0,dpdown:b16,dpright:+a0,dpleft:-a0,dpup:h0.1,leftx:a0,lefty:a1,righty:a3~,lefttrigger:a2,'
# ---
# name: TestController.test_generate_sdl_game_db_line_ignore_buttons
  '030000005e040000a102000007010000,real name,platform:Linux,b:b1,a:b0,dpdown:b16,dpright:+a0,dpleft:-a0,dpup:h0.1,leftx:a0,lefty:a1,righty:a3~,lefttrigger:a2,'
# ---
# name: TestController.test_generate_sdl_game_db_line_no_hotkey
  '030000005e040000a102000007010000,real name,platform:Linux,b:b1,a:b0,dpdown:b16,dpright:+a0,dpleft:-a0,dpup:h0.1,leftx:a0,lefty:a1,righty:a3~,lefttrigger:a2,'
# ---
# name: TestController.test_init
  Controller(name='name', type='joystick', guid='guid', player_number=1, index=0, real_name='real name', device_path='/dev/null', button_count=1, hat_count=2, axis_count=3, physical_device_path=None, physical_index=None, inputs={'a': Input(name='a', type='button', id='1', value='1', code='305')})
# ---
# name: TestController.test_load_for_players
  list([
    Controller(name='Test Controller', type='joystick', guid='2', player_number=1, index=0, real_name='Test Controller', device_path='/dev/event/input1', button_count=1, hat_count=2, axis_count=3, physical_device_path=None, physical_index=None, inputs={'b': Input(name='b', type='button', id='0', value='1', code='222')}),
    Controller(name='Test Controller', type='joystick', guid='2', player_number=2, index=22, real_name='Something else', device_path='/dev/event/input2', button_count=1, hat_count=2, axis_count=3, physical_device_path=None, physical_index=None, inputs={'b': Input(name='b', type='button', id='0', value='1', code='222')}),
    Controller(name='Test Controller', type='joystick', guid='20', player_number=3, index=2, real_name='Test Controller', device_path='/dev/event/input3', button_count=1, hat_count=2, axis_count=3, physical_device_path=None, physical_index=None, inputs={'b': Input(name='b', type='button', id='0', value='1', code='222')}),
    Controller(name='Test Controller', type='joystick', guid='1', player_number=4, index=4, real_name='Test Controller', device_path='/dev/event/input4', button_count=1, hat_count=2, axis_count=3, physical_device_path=None, physical_index=None, inputs={'a': Input(name='a', type='button', id='0', value='1', code='111')}),
    Controller(name='Test Controller 2', type='joystick', guid='3', player_number=5, index=6, real_name='Test controller 2', device_path='/dev/event/input6', button_count=1, hat_count=2, axis_count=3, physical_device_path=None, physical_index=None, inputs={'x': Input(name='x', type='button', id='0', value='1', code='111')}),
  ])
# ---
# name: TestController.test_load_for_players_no_user_config
  list([
    Controller(name='Test Controller', type='joystick', guid='1', player_number=1, index=0, real_name='Test Controller', device_path='/dev/event/input1', button_count=1, hat_count=2, axis_count=3, physical_device_path=None, physical_index=None, inputs={'a': Input(name='a', type='button', id='0', value='1', code='111')}),
    Controller(name='Test Controller', type='joystick', guid='1', player_number=2, index=22, real_name='Something else', device_path='/dev/event/input2', button_count=1, hat_count=2, axis_count=3, physical_device_path=None, physical_index=None, inputs={'a': Input(name='a', type='button', id='0', value='1', code='111')}),
    Controller(name='Test Controller', type='joystick', guid='20', player_number=3, index=2, real_name='Test Controller', device_path='/dev/event/input3', button_count=1, hat_count=2, axis_count=3, physical_device_path=None, physical_index=None, inputs={'a': Input(name='a', type='button', id='0', value='1', code='111')}),
    Controller(name='Test Controller', type='joystick', guid='1', player_number=4, index=4, real_name='Test Controller', device_path='/dev/event/input4', button_count=1, hat_count=2, axis_count=3, physical_device_path=None, physical_index=None, inputs={'a': Input(name='a', type='button', id='0', value='1', code='111')}),
    Controller(name='Test Controller 2', type='joystick', guid='3', player_number=5, index=6, real_name='Test controller 2', device_path='/dev/event/input6', button_count=1, hat_count=2, axis_count=3, physical_device_path=None, physical_index=None, inputs={'x': Input(name='x', type='button', id='0', value='1', code='111')}),
  ])
# ---
